Source/SIMPLib/Messages/AbstractMessageHandler.h

```diff
@@ -0,0 +1,80 @@
+* other materials provided with the distribution.
+*
+* Neither the name of BlueQuartz Software, the US Air Force, nor the names of its
+* contributors may be used to endorse or promote products derived from this software
+* without specific prior written permission.
+*
+* THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+* AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+* IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+* DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+* FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+* DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+* SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+* CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+* OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+* USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*
+* The code contained herein was partially funded by the followig contracts:
+*    United States Air Force Prime Contract FA8650-07-D-5800
+*    United States Air Force Prime Contract FA8650-10-D-5210
+*    United States Prime Contract Navy N00173-07-C-2068
+*
+* ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ */
+
+#pragma once
+
+#include "SIMPLib/SIMPLib.h"
+
+class GenericErrorMessage;
+class GenericProgressMessage;
+class GenericStatusMessage;
+class GenericWarningMessage;
+class PipelineErrorMessage;
+class PipelineProgressMessage;
+class PipelineStatusMessage;
+class PipelineWarningMessage;
+class FilterErrorMessage;
+class FilterProgressMessage;
+class FilterStatusMessage;
+class FilterWarningMessage;
+
+/**
+ * @class AbstractMessageHandler AbstractMessageHandler.h SIMPLib/Messages/AbstractMessageHandler.h
+ * @brief This is the message handler superclass that is part of the double-dispatch API and enables
+ * observers to process messages that they receive from observable objects
+ */
+class SIMPLib_EXPORT AbstractMessageHandler
+{
+  public:
+    virtual void processMessage(GenericErrorMessage* msg) const;
+    virtual void processMessage(GenericProgressMessage* msg) const;
+    virtual void processMessage(GenericStatusMessage* msg) const;
+    virtual void processMessage(GenericWarningMessage* msg) const;
+
+    virtual void processMessage(PipelineErrorMessage* msg) const;
+    virtual void processMessage(PipelineProgressMessage* msg) const;
+    virtual void processMessage(PipelineStatusMessage* msg) const;
+    virtual void processMessage(PipelineWarningMessage* msg) const;
+
+    virtual void processMessage(FilterErrorMessage* msg) const;
+    virtual void processMessage(FilterProgressMessage* msg) const;
+    virtual void processMessage(FilterStatusMessage* msg) const;
+    virtual void processMessage(FilterWarningMessage* msg) const;
```

Comment:
Is there isn't a case where we should be editing the messages in message handler?  If not, these should be const parameters.  My understanding so far is that we should not be changing the message after it has been created.

Reply:
From the name of the method is it "ProcessMessage" with no indication that any editing of the object would occur so make the arguments into const FilterErrorMessage* msg. Also add DOxygen documentation for these methods to amply describe what they are doing.